plutus-core/testlib/PlutusCore/Generators/PIR/GenTm.hs

```diff
@@ -0,0 +1,305 @@
+-- in the second.
+sizeSplit :: Int -> Int -> GenTm a -> (a -> GenTm b) -> GenTm (a, b)
+sizeSplit a b ga gb = do
+  n <- asks geSize
+  let na = (a * n) `div` (a + b)
+      nb = (b * n) `div` (a + b)
+  x <- withSize na ga
+  (,) x <$> withSize nb (gb x)
+
+-- * Names
+
+-- | Get all uniques we have generated and are used in the current context.
+getUniques :: GenTm (Set Unique)
+getUniques = do
+  GenEnv{geDatas = dts, geTypes = tys, geTerms = tms, geUnboundUsedTyNames = used} <- ask
+  return $ Set.mapMonotonic (_nameUnique . unTyName) (Map.keysSet dts <> Map.keysSet tys <> used) <>
+           Set.mapMonotonic _nameUnique (Map.keysSet tms) <>
+           Set.unions [ names d | d <- Map.elems dts ]
+  where
+    names (Datatype _ _ _ m cs) = Set.fromList $ _nameUnique m : [ _nameUnique c | VarDecl _ c _ <- cs ]
+
+{- Note [Warning about generating fresh names]: because `GenTm` is a *reader* monad
+   names are not immediately put into any state when generated. There is *no guarantee*
+   that in this situation:
+   ```
+   do nms <- genFreshNames ss
+      nms' <- genFreshNames ss
+   ```
+   the names in `nms` and `nms'` don't overlap.
+
+   Instead, what you are supposed to do is locally use the names in `nms` and `nms'` to
+   define generators that use them. This is done with functions like `bindTyName` and `bindTmName`:
+   ```
+   genLam ma mb = do
+      x <- genFreshName "x"
+      sizeSplit 1 7 (maybe (genType Star) return ma)
+                    --      v--- LOOK HERE!
+                    (\ a -> bindTmName x a . noEscape $ genTerm mb) $ \ a (b, body) ->
+                    --      ^--- LOOK HERE!
+                    TyFun () a b, LamAbs () x a body)
+   ```
+-}
+
+-- | Generate a fresh name. See Note [Warning about generating fresh names].
+genFreshName :: String -> GenTm Name
+genFreshName s = head <$> genFreshNames [s]
+
+-- | Generate one fresh name per string in the input list.
+-- names don't overlap. See Note [Warning about generating fresh names].
+genFreshNames :: [String] -> GenTm [Name]
+genFreshNames ss = do
+  used <- getUniques
+  let i = fromEnum $ Set.findMax $ Set.insert (Unique 0) used
+      js = [ j | j <- [1..i], not $ Unique j `Set.member` used ]
+      is = js ++ take (length ss + 10) [i+1..]
+  is' <- liftGen $ QC.shuffle is
+  return [Name (fromString $ s ++ show j) (toEnum j) | (s, j) <- zip ss is']
+
+-- | See `genFreshName`
+genFreshTyName :: String -> GenTm TyName
+genFreshTyName s = TyName <$> genFreshName s
+
+-- | See `genFreshNames`
+genFreshTyNames :: [String] -> GenTm [TyName]
+genFreshTyNames ss = map TyName <$> genFreshNames ss
+
+-- | Generate a name that overlaps with existing names on purpose. If there
+-- are no existing names, generate a fresh name.
+genNotFreshName :: String -> GenTm Name
+genNotFreshName s = do
+  used <- Set.toList <$> getUniques
+  case used of
+    [] -> genFreshName s
+    _  -> liftGen $ elements [ Name (fromString $ s ++ show (unUnique i)) i | i <- used ]
```

Comment:
That doesn't make the function incorrect, only incomplete. Which one may want to address by splitting uniques for types and terms.

Reply:
> That doesn't make the function incorrect, only incomplete. 

Yeah, I mean, the name and the docs suggest that the generated name is supposed not to be fresh, while in certain cases it can. Anyways, ignore my nitpicking, I'll just clarify the docs a little.